sync15-adapter/src/changeset.rs

```diff
@@ -39,27 +40,38 @@ impl<T> RecordChangeset<T> {
 
 impl OutgoingChangeset {
     pub fn encrypt(self, key: &KeyBundle) -> Result<Vec<EncryptedBso>> {
-        let RecordChangeset { changes, collection, .. } = self;
-        changes.into_iter()
-               .map(|change| change.into_bso(collection.clone()).encrypt(key))
-               .collect()
+        let RecordChangeset {
+            changes,
+            collection,
+            ..
+        } = self;
+        changes
+            .into_iter()
+            .map(|change| change.into_bso(collection.clone()).encrypt(key))
+            .collect()
     }
 
-    pub fn post(self, svc: &Sync15Service, fully_atomic: bool) -> Result<UploadInfo> {
-        Ok(CollectionUpdate::new_from_changeset(svc, self, fully_atomic)?.upload()?)
+    pub fn post(
+        self,
+        client: &Sync15StorageClient,
+        state: &GlobalState,
+        fully_atomic: bool,
+    ) -> Result<UploadInfo> {
+        Ok(CollectionUpdate::new_from_changeset(client, state, self, fully_atomic)?.upload()?)
     }
 }
 
 impl IncomingChangeset {
     pub fn fetch(
-        svc: &Sync15Service,
+        client: &Sync15StorageClient,
+        state: &GlobalState,
         collection: String,
-        since: ServerTimestamp
+        since: ServerTimestamp,
     ) -> Result<IncomingChangeset> {
-        let records = svc.get_encrypted_records(&collection, since)?;
-        let mut result = IncomingChangeset::new(collection, svc.last_server_time());
+        let records = client.get_encrypted_records(&collection, since)?;
+        let mut result = IncomingChangeset::new(collection, client.last_server_time());
```

Comment:
It seems like this is wrong (or at least misleading) - but it was before too, so really I'm just mentioning it.

IIUC, it will be used as the XIUS header if the first request is a POST - but in that case it should be the collection timestamp. However, that should also never happen - we should always do a fetch first - and the fetch method take the timestamp as a param - and it's not clear to me that this is correctly updated after a fetch.

I wonder if we should initialize this to an "impossible" value (eg, 0)?

Reply:
I was wondering about that, too. If I'm reading [this](https://github.com/mozilla-services/server-syncstorage/blob/1ec5758be224c5fd47d54101a35725df2dbe9ea4/syncstorage/views/util.py#L62-L68) and [this](https://github.com/mozilla-services/server-syncstorage/blob/1ec5758be224c5fd47d54101a35725df2dbe9ea4/syncstorage/views/decorators.py#L156-L162) right, yes, the server checks against the collection timestamp, not the current time that it returns in `X-Weave-Timestamp`. (We shouldn't get a 412 for [our own writes](https://github.com/mozilla-services/server-syncstorage/blob/1ec5758be224c5fd47d54101a35725df2dbe9ea4/syncstorage/tweens.py#L31-L34), but we might send a newer timestamp that might not be correct if another client synced in the meantime). Since we store `i/c` in `GlobalState`, we could probably just use `state.last_modified_or_zero(&collection)` here.

Reply:
Your description matches my expectation of server behaviour; `X-Weave-Timestamp` is mostly a legacy thing and the server tries hard to avoid using it, preferring more meaningful timestamps like `X-Last-Modified`.